desugarer.go

```diff
@@ -16,15 +16,387 @@ limitations under the License.
+		case '\\':
+			if i >= len(s) {
+				return "", makeStaticError("Truncated escape sequence in string literal.", *loc)
+			}
+			r2, w := utf8.DecodeRuneInString(s[i:])
+			i += w
+			switch r2 {
+			case '"':
+				buf.WriteRune('"')
+			case '\\':
+				buf.WriteRune('\\')
+			case '/':
+				buf.WriteRune('/') // This one is odd, maybe a mistake.
+			case 'b':
+				buf.WriteRune('\b')
+			case 'f':
+				buf.WriteRune('\f')
+			case 'n':
+				buf.WriteRune('\n')
+			case 'r':
+				buf.WriteRune('\r')
+			case 't':
+				buf.WriteRune('\t')
+			case 'u':
+				if i+4 > len(s) {
+					return "", makeStaticError("Truncated unicode escape sequence in string literal.", *loc)
+				}
+				codeBytes, err := hex.DecodeString(s[0:4])
+				if err != nil {
+					return "", makeStaticError(fmt.Sprintf("Unicode escape sequence was malformed: %s", s[0:4]), *loc)
+				}
+				code := int(codeBytes[0])*256 + int(codeBytes[1])
+				buf.WriteRune(rune(code))
+				i += 4
+			default:
+				return "", makeStaticError(fmt.Sprintf("Unknown escape sequence in string literal: \\%c", r2), *loc)
+			}
+
+		default:
+			buf.WriteRune(r)
+		}
+	}
+	return buf.String(), nil
+}
+
+func desugarFields(location LocationRange, fields *astObjectFields, objLevel int) error {
+
+	// Desugar children
+	for _, field := range *fields {
+		if field.expr1 != nil {
+			err := desugar(&field.expr1, objLevel)
+			if err != nil {
+				return err
+			}
+		}
+		err := desugar(&field.expr2, objLevel+1)
+		if err != nil {
+			return err
+		}
+		if field.expr3 != nil {
+			err := desugar(&field.expr3, objLevel+1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Simplify asserts
+	// TODO(dcunnin): this
+	for _, field := range *fields {
+		if field.kind != astObjectAssert {
+			continue
+		}
+		/*
+			AST *msg = field.expr3
+			field.expr3 = nil
+			if (msg == nil) {
+				auto msg_str = U"Object assertion failed."
+				msg = alloc->make<LiteralString>(field.expr2->location, msg_str,
+												 LiteralString::DOUBLE, "")
+			}
+
+			// if expr2 then true else error msg
+			field.expr2 = alloc->make<Conditional>(
+				ast->location,
+				field.expr2,
+				alloc->make<LiteralBoolean>(E, true),
+				alloc->make<Error>(msg->location, msg))
+		*/
+	}
+
+	// Remove methods
+	// TODO(dcunnin): this
+	for _, field := range *fields {
+		if !field.methodSugar {
+			continue
+		}
+		/*
+			field.expr2 = alloc->make<Function>(
+				field.expr2->location, field.ids, false, field.expr2)
+			field.methodSugar = false
+			field.ids.clear()
+		*/
+	}
+
+	// Remove object-level locals
+	newFields := []astObjectField{}
+	var binds astLocalBinds
+	for _, local := range *fields {
+		if local.kind != astObjectLocal {
+			continue
+		}
+		binds = append(binds, astLocalBind{variable: *local.id, body: local.expr2})
+	}
+	for _, field := range *fields {
+		if field.kind == astObjectLocal {
+			continue
+		}
+		if len(binds) > 0 {
+			field.expr2 = &astLocal{astNodeBase{loc: *field.expr2.Loc()}, binds, field.expr2}
```

Comment:
Do we want to create astLocal once and share it across all of the fields?  Are there places where we mutate it later?